pkg/controller/runtime/adapter.go

```diff
@@ -479,3 +499,52 @@ func (adapter *adapter) runOnce(ctx context.Context, logger *zap.Logger) (err er
 
 	return adapter.ctrl.Run(ctx, adapter, logger)
 }
+
+func outputTrackingID(md *resource.Metadata) string {
+	return fmt.Sprintf("%s/%s/%s", md.Namespace(), md.Type(), md.ID())
+}
```

Comment:
Why string tho? Can't we return custom struct or use `ordered.Triple` from `gen`? This will remove unnecessary allocations

Reply:
good idea, I will rework

Reply:
updated